net: dns: Allow using resolver and responder at the same time

jukkar opened the pull request:
This allows user to have both mDNS resolver and responder enabled at the same time.

Fixes #72553

jukkar committed:
net: dns: Allow using resolver and responder at the same time

Allow mDNS resolver and responder to to be used at the same
time so that both can use the port 5353. This requires
a DNS traffic dispatcher which affects also the normal DNS
resolver.

Fixes #72553

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Start socket service thread by net core init

Do not depend on init level but start the socket service
already in net core init because DNS init code depends on
socket service API to be ready to serve. And we call DNS
init at the net core init.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Add ANY query resource type

Don't give an error for ANY type record.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Add helper for figuring out the query type

Add helper function that returns the name of the query type
so that we can print it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: llmnr_responder: Print query type properly

Instead of printing either A or AAAA resource query type,
print the correct query type value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
@besmarsh, I have a fix proposal for the mDNS resolver/responder co-existence issue you created. Would you be able to try it and report if it works for you or not?

besmarsh commented:
> @besmarsh, I have a fix proposal for the mDNS resolver/responder co-existence issue you created. Would you be able to try it and report if it works for you or not?

For sure, I should be able to test it in the next couple of days.

jukkar committed:
samples: net: Increase CONFIG_NET_SOCKETS_POLL_MAX to pass checks

The number of suitable network IPv6/4 configurations in interfaces
affect the number of pollable sockets. So need to increase the value
from the default value of 3.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/dns/dispatcher.c:
```diff
@@ -0,0 +1,304 @@
+	dns_data = net_buf_alloc(&dns_msg_pool, dispatcher->buf_timeout);
+	if (!dns_data) {
+		ret = DNS_EAI_MEMORY;
+		goto free_buf;
```
It's rather harmless given there's a pointer check under `free_buf`, but since the allocation failed we could jump directly to `unlock` label here.

rlubos commented on subsys/net/lib/dns/dispatcher.c:
```diff
@@ -0,0 +1,304 @@
+							   dns_data, data_len);
+			} else {
+				/* The pair must be a resolver */
+				CHECKIF(dispatcher->pair->type != DNS_SOCKET_RESOLVER) {
```
Perhaps we should also check for errors like this in `dns_dispatcher_register()`? So that you don't need the actual communication going on to figure out something is wrong.

jukkar replied:
Indeed, good point, will change this.

jukkar replied:
Actually we already check in `dns_dispatcher_register()` that the dispatcher type is not the same, so this checkif is not really needed.

rlubos reviewed (commented):
Looks pretty good

jukkar commented:
* While testing, I noticed an issue when wrong context was used in dispatcher callback

besmarsh reviewed (commented):
Tested and confirmed to work with my custom application. I've left some minor code review comments. Thanks @jukkar for the quick turnaround on this fix!

jukkar commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -845,32 +776,33 @@ static int init_listener(void)
-				"IPv4", MAX_IPV4_IFACE_COUNT, CONFIG_NET_MAX_CONTEXTS);
+			NET_ERR("Cannot get %s socket (%d %s interfaces). Max sockets is %d",
+				"IPv4", MAX_IPV4_IFACE_COUNT,
+				"IPv4", CONFIG_NET_MAX_CONTEXTS);
```
Yes, we avoid having two string stored to flash by sharing the general format for it.

jukkar commented:
* Updated according to comments
